<a href="https://colab.research.google.com/github/CppR-Repo/TT-2C2025-Data-Analitycs-Notebooks/blob/main/PreEntrega/MiPre-entrega/Pre-entrega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Análisis de datos con Pyhon - Pre-entrega
Nombre y apellido: Cristian Pablo Portillo Rios

# Etapa 1

## Script básico
Actividades 2 y 3

In [3]:
# Generamos una lista de diccionarios que contenga algunos datos
ventas = [
  {'producto': 'casablanca', 'tipo': 'latex', 'descripcion': 20, 'precio': 120000.0, 'cantidad': 2},
  {'producto': 'satinol', 'tipo': 'sintetico', 'descripcion': 4, 'precio': 75000.0, 'cantidad': 3},
  {'producto': 'andina', 'tipo': 'fijador', 'descripcion': 1, 'precio': 7800.0, 'cantidad': 2},
  {'producto': 'comodin', 'tipo': 'enduido', 'descripcion': 10, 'precio': 20500.0, 'cantidad': 1},
  {'producto': 'aa', 'tipo': 'lija', 'descripcion': 100, 'precio': 2100.0, 'cantidad': 10}
]

In [4]:
# Calculamos las ventas acumuladas - map
# 1. Calcula la venta de cada producto con map
ventas_individuales = map(lambda x:x['precio'] * x['cantidad'], ventas)

# 2. Sumar todas las ventas individuales con sum
ventas_acumuladas = sum(ventas_individuales)
print(f"Las ventas acumuladas son: {ventas_acumuladas}")

Las ventas acumuladas son: 522100.0


In [5]:
# Calculamos las ventas acumuladas - listas por comprensión
venta_acumuladas = sum([venta["precio"] * venta["cantidad"] for venta in ventas])
venta_acumuladas

522100.0

## Import Datasets
Actividad 1

In [6]:
import pandas as pd

In [7]:
# Import dataset ventas
df_ventas = pd.read_csv("https://raw.githubusercontent.com/CppR-Repo/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/ventas.csv")

In [8]:
# Import dataset clientes
df_clientes = pd.read_csv("https://raw.githubusercontent.com/CppR-Repo/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/clientes.csv")

In [9]:
# Import dataset marketing
df_marketing = pd.read_csv("https://raw.githubusercontent.com/CppR-Repo/TT-2C2025-Data-Analitycs-Notebooks/refs/heads/main/PreEntrega/Datasets/marketing.csv")

In [10]:
df_ventas

id_venta                producto   precio  cantidad fecha_venta  \
0          792       Cuadro decorativo   $69.94       5.0  02/01/2024   
1          811         Lámpara de mesa  $105.10       5.0  02/01/2024   
2         1156                Secadora   $97.96       3.0  02/01/2024   
3         1372                Heladera  $114.35       8.0  02/01/2024   
4         1546                Secadora  $106.21       4.0  02/01/2024   
...        ...                     ...      ...       ...         ...   
3030      1837         Horno eléctrico  $104.12       9.0  30/12/2024   
3031      2276                  Laptop   $85.27       9.0  30/12/2024   
3032      2696                  Laptop  $107.81       4.0  30/12/2024   
3033      2913              Smartphone   $99.85       7.0  30/12/2024   
3034      2930  Consola de videojuegos   $55.47       6.0  30/12/2024   

              categoria  
0            Decoración  
1            Decoración  
2     Electrodomésticos  
3     Electrodomésticos  
4     Electrodomésticos  
...                 ...  
3030  Electrodomésticos  
3031        Electrónica  
3032        Electrónica  
3033        Electrónica  
3034        Electrónica  

[3035 rows x 6 columns]

## Dataframe ventas

### Análisis exploratorio de **datos**
Actividad 4

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [11]:
# Aplicamos info() para familiarizarnos con el dataframe
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


In [12]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_ventas.head()

id_venta           producto   precio  cantidad fecha_venta  \
0       792  Cuadro decorativo   $69.94       5.0  02/01/2024   
1       811    Lámpara de mesa  $105.10       5.0  02/01/2024   
2      1156           Secadora   $97.96       3.0  02/01/2024   
3      1372           Heladera  $114.35       8.0  02/01/2024   
4      1546           Secadora  $106.21       4.0  02/01/2024   

           categoria  
0         Decoración  
1         Decoración  
2  Electrodomésticos  
3  Electrodomésticos  
4  Electrodomésticos

In [13]:
# Usar sample() para tomar registros al azar
df_ventas.sample(5)

id_venta          producto   precio  cantidad fecha_venta  \
1881      2549         Proyector   $69.11      11.0  05/08/2024   
1081      2367    Cámara digital   $91.04       8.0  29/04/2024   
693        950   Adorno de pared  $116.42       6.0  20/03/2024   
1629      1770  Plancha de vapor   $80.83       9.0  06/07/2024   
130       2477    Cámara digital   $38.37       9.0  16/01/2024   

              categoria  
1881        Electrónica  
1081        Electrónica  
693          Decoración  
1629  Electrodomésticos  
130         Electrónica

In [14]:
# Analizamos el campo precio
# Si es numericos se puede aplicar describe()
df_ventas["precio"].sample(5)

998     $119.07
2581     $85.31
2432     $85.98
2244     $67.94
192     $102.40
Name: precio, dtype: object

In [15]:
# Analizamos el campo cantidad
# si es numerico se puede aplicar describe()
df_ventas["cantidad"].sample(5)

354     12.0
481      1.0
2890     4.0
1662     9.0
1692     4.0
Name: cantidad, dtype: float64

In [16]:
# Analizamos el campo fecha_venta
df_ventas["fecha_venta"].sample(5)

3027    30/12/2024
1998    22/08/2024
1060    27/04/2024
1961    17/08/2024
1339    28/05/2024
Name: fecha_venta, dtype: object

In [17]:
# Analizamos el campo categoria
# Probar con value_counts y unique
df_ventas["categoria"].value_counts()

categoria
Decoración           1015
Electrónica          1012
Electrodomésticos    1008
Name: count, dtype: int64

### Calidad de datos
Actividad 5

In [19]:
# Buscamos registros duplicados
df_ventas.duplicated().sum()

np.int64(35)

In [21]:
# Visualizamos los duplicados
# keep: 'first', 'last', False
df_ventas[df_ventas.duplicated(keep='first')]

id_venta                producto   precio  cantidad fecha_venta  \
826        56                Cortinas   $66.24       5.0  05/04/2024   
827       421         Lámpara de mesa  $114.83       9.0  05/04/2024   
828       424       Jarrón decorativo   $87.94       2.0  05/04/2024   
829      1868                Cafetera   $62.23       1.0  05/04/2024   
830      2545             Auriculares   $32.81      11.0  05/04/2024   
831      2778              SmartWatch   $70.66       7.0  05/04/2024   
844       145                Alfombra   $41.72       8.0  06/04/2024   
845       300         Adorno de pared   $67.75       9.0  06/04/2024   
846       439         Lámpara de mesa   $35.43       2.0  06/04/2024   
847       906                Cortinas   $84.04       3.0  06/04/2024   
848       959         Lámpara de mesa   $76.32       7.0  06/04/2024   
849      1376                Secadora   $69.58       7.0  06/04/2024   
850      1776                Secadora   $33.55       3.0  06/04/2024   
851      1784              Aspiradora  $115.24       3.0  06/04/2024   
852      1974              Aspiradora   $87.32       7.0  06/04/2024   
853      2080  Consola de videojuegos  $100.69       7.0  06/04/2024   
854      2108              SmartWatch   $79.37      10.0  06/04/2024   
855      2169               Proyector  $120.34       1.0  06/04/2024   
867       213       Espejo decorativo   $40.76       8.0  07/04/2024   
868       345                Alfombra   $56.78       9.0  07/04/2024   
869       398       Rincón de plantas   $77.30       7.0  07/04/2024   
870       509         Lámpara de mesa   $72.50       1.0  07/04/2024   
871       821         Lámpara de mesa   $71.17      10.0  07/04/2024   
872      1642                Heladera   $88.78       2.0  07/04/2024   
873      1956                Secadora   $42.59      12.0  07/04/2024   
874      2085             Auriculares  $122.56       6.0  07/04/2024   
875      2096                  Laptop   $83.25       6.0  07/04/2024   
876      2356                  Laptop   $30.16       6.0  07/04/2024   
877      2373              Smartphone  $123.60      12.0  07/04/2024   
884       204       Jarrón decorativo   $94.75       1.0  08/04/2024   
885      1381      Freidora eléctrica   $38.12       2.0  08/04/2024   
886      2365             Auriculares   $92.91      11.0  08/04/2024   
887      2486                  Laptop  $124.95      11.0  08/04/2024   
888      2506                  Laptop   $34.74       1.0  08/04/2024   
889      2705             Auriculares   $87.29       2.0  08/04/2024   

             categoria  
826         Decoración  
827         Decoración  
828         Decoración  
829  Electrodomésticos  
830        Electrónica  
831        Electrónica  
844         Decoración  
845         Decoración  
846         Decoración  
847         Decoración  
848         Decoración  
849  Electrodomésticos  
850  Electrodomésticos  
851  Electrodomésticos  
852  Electrodomésticos  
853        Electrónica  
854        Electrónica  
855        Electrónica  
867         Decoración  
868         Decoración  
869         Decoración  
870         Decoración  
871         Decoración  
872  Electrodomésticos  
873  Electrodomésticos  
874        Electrónica  
875        Electrónica  
876        Electrónica  
877        Electrónica  
884         Decoración  
885  Electrodomésticos  
886        Electrónica  
887        Electrónica  
888        Electrónica  
889        Electrónica

In [22]:
# Buscamos datos nulos
df_ventas.isnull().sum()

id_venta       0
producto       0
precio         2
cantidad       2
fecha_venta    0
categoria      0
dtype: int64

In [23]:
# Visualizar los nulos
# df_ventas[df_ventas.isnull().any(axis=1)]
df_ventas[df_ventas.isnull().any(axis=1)]

id_venta               producto precio  cantidad fecha_venta    categoria
136       627  Elementos de cerámica    NaN       NaN  17/01/2024   Decoración
139      2171    Parlantes Bluetooth    NaN       NaN  17/01/2024  Electrónica

In [24]:
# Analizar los precios de los NaN a ver que opciones tenemos
# df_ventas[df_ventas["producto"]=="Elementos de cerámica"]
df_ventas["producto"].value_counts()
# Unique nos devuelve una lista de los valores unicos.
# df_ventas["producto"].unique()


producto
Lámpara de mesa           181
Auriculares               147
Microondas                135
Cafetera                  118
Laptop                    104
Secadora                  103
Jarrón decorativo         102
Smartphone                102
Alfombra                  102
SmartWatch                102
Aspiradora                102
Cortinas                  102
Rincón de plantas         102
Freidora eléctrica        101
Espejo decorativo         101
Adorno de pared           101
Lavadora                  100
Consola de videojuegos    100
Cuadro decorativo         100
Horno eléctrico           100
Plancha de vapor          100
Proyector                 100
Elementos de cerámica     100
Batidora                  100
Televisor                 100
Cámara digital            100
Parlantes Bluetooth        99
Tablet                     58
Heladera                   49
Candelabro                 24
Name: count, dtype: int64

### Observaciones




* Cantidad de registros: 3035
* Columnas 6
* 35 duplicados, eliminarlos
* Algunos nulos (2), 1% del faltante, dropna
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64
* Campo fecha: ajustar a dtype datetime

## Dataframe marketing

### Analisis exploratorio de datos

Aplicamos los métodos de EDA vistos en clase, `info`, `columns`, `dtypes`, `head`, `tail`, `sample`, `unique`, `value_counts`, `describe`, `duplicated`, `isnull`, etc. y documentamos las observaciones:

In [ ]:
# Aplicamos info() para familiarizarnos con el dataframe
df_marketing.info()

In [ ]:
# Visualizamos los primeros registros con head()
# también podemos usar tail() para visualizar los últimos
df_marketing.head()

In [ ]:
# Analizamos el campo costo
df_marketing["costo"].sample(5)

In [ ]:
# Analizamos el campo fecha_inicio
df_marketing["fecha_inicio"].sample(5)

In [ ]:
# Analizamos el campo fecha_fin
df_marketing["fecha_fin"].sample(5)

### Calidad de los datos


In [ ]:
# Buscamos registros duplicados


### Observaciones
*** Completar ****

## Dataframe clientes

### Analisis exploratorio de datos

### Observaciones

# Etapa 2

## Limpieza de datos
Actividad 1 - Limpiar el conjunto de datos eliminando duplicados y caracteres no deseados. Documentar el proceso y los resultados.

### Dataframe ventas

* 35 duplicados, eliminarlos - se eliminaron 35 duplicados
* Algunos nulos (2), 1% del faltante, dropna - se eliminaron 2 registros con 2 NaN
* Campo precio: encontre `$`, removerlo y luego cambiar el dtype a float64 - Ok
* Campo fecha: ajustar a dtype datetime - Ok

In [ ]:
# Eliminamos registros duplicados
# atributos a evaluar, keep, ignore_index, inplace
df_ventas.drop_duplicates(inplace=True, ignore_index=True)

In [ ]:
# Validamos que no haya duplicados
df_ventas.duplicated().sum()

In [ ]:
# Eliminamos registros con datos nulos
df_ventas.dropna(inplace=True)

In [ ]:
# Validamos que no haya nulos
# df_ventas.isnull().sum()
df_ventas.info()

In [ ]:
# Eliminamos el caracter especial $ de precio
# Normalización: usar replace()
# De ser necesario, podemos usar el atributo regex=True
df_ventas["precio"] = df_ventas["precio"].str.strip().str.replace("$", "")

In [ ]:
# Validamos que precio no contenga $
df_ventas["precio"].sample(3)

In [ ]:
# Convertimos precio de str a float
# Podemos usar pd.to_numeric o astype o incluso combinarlos
df_ventas["precio"] = df_ventas["precio"].astype(float)

In [ ]:
# Convertir el dtype de cantidad, de float64 a int64
df_ventas["cantidad"]= df_ventas["cantidad"].astype(int)

In [ ]:
# Validar la conversion a int
df_ventas["cantidad"].dtypes

In [ ]:
# Convertir el campo fecha_venta a dtype datetime
# pd.to_datetime(df_ventas["fecha_venta"], format="%d/%m/%Y")
# pd.to_datetime(df_ventas["fecha_venta"], dayfirst=True)
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format="%d/%m/%Y")

In [ ]:
# Validar la conversion de fecha
df_ventas['fecha_venta'].dtype

### Dataframe marketing

*** Acciones ***

In [ ]:
# Convertir el campo fecha_inicio a dtype datetime


In [ ]:
# Convertir el campo fecha_fin a dtype datetime


In [ ]:
# Validamos los cambios


## Transformación de datos
Actividad 2 - Aplicar filtros y transformaciones para crear una tabla de ventas que muestre solo los productos con alto rendimiento.

In [ ]:
df_ventas.head(2)

In [ ]:
# Transformación:
# Calculamos el valor de cada venta = precio * cantidad
# Agregamos una nueva columnas
df_ventas["valor_venta"] = df_ventas["precio"] * df_ventas["cantidad"]

In [ ]:
# Visualizamos
df_ventas.head(2)

In [ ]:
# Agregación:
# Agrupar valor_venta por producto
# Sumamos las ventas dentro de cada grupo
tabla_ventas = (df_ventas.groupby("producto", as_index=False)["valor_venta"].sum())
tabla_ventas.head()

In [ ]:
# Usemos describe para analizar algunos valores del dataframe resultante
tabla_ventas.describe()

In [ ]:
# Agregamos un plot para visualizar graficamente la distribución
import seaborn as sns
import matplotlib.pyplot as plt

sns.histplot(tabla_ventas["valor_venta"], bins=30, kde=True, edgecolor="black")
plt.title("Distribución de Valor de Venta con Curva de Densidad")
plt.xlabel("Valor de venta")
plt.ylabel("Frecuencia")
plt.show()

In [ ]:
# Calcular el percentil 75 (umbral de alto rendimiento)
umbral_ventas = tabla_ventas["valor_venta"].quantile(0.75)
umbral_ventas

In [ ]:
# Identificar productos que superan el umbral
# productos_top = tabla_ventas.query("valor_venta >= @umbral")
productos_top = tabla_ventas[tabla_ventas["valor_venta"] >= umbral_ventas]
productos_top

## Agregación
Actividad 3 - Resumir las ventas por categoría de producto y analizar los ingresos generados.

In [ ]:
df_ventas.columns

In [ ]:
# Agrupamos por categoria y agregamos la suma
ventas_categoria = df_ventas.groupby("categoria", as_index=False)["valor_venta"].sum()
ventas_categoria.head()

In [ ]:
ventas_categoria

In [ ]:
# Ordenar y mostrar los resultados
ventas_categoria

## Integración de datos
Actividad 4 - Combinar los sets de datos de ventas y marketing para obtener una visión más amplia de las tendencias.

In [ ]:
df_ventas.head(2)

In [ ]:
df_marketing.info()

In [ ]:
# Unir los datasets ventas y marketing por columna 'producto'
ventas_marketing = pd.merge(df_ventas, df_marketing, on="producto", how="left")

In [ ]:
# Verificar coincidencias (muestra)
ventas_marketing.head()

### Analizamos el impacto de la campaña

In [ ]:
# Nos quedamos solo con los registros de ventas durante la campaña de marketing
# Seleccionamos solo los campos que nos interesan: "producto", "valor_venta", "costo"


In [ ]:
# Agrupamos por producto y hacemos una agregación por "valor de venta" y "costo"
ventas_marketing.groupby("categoria")[["valor_venta","costo"]].sum()

In [ ]:
# Generamos una nueva columna con la diferencia entre lo vendido y lo invertido en la campaña
